java/Xpedite/src/com/xpedite/ClassTransformer.java

```diff
@@ -0,0 +1,70 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Custom class transformer to transform Java classes inserting Xpedite probes
+//
+// Author: Brooke Elizabeth Cantwell, Morgan Stanley
+//
+///////////////////////////////////////////////////////////////////////////////
+
+package com.xpedite;
+import com.xpedite.probes.AbstractProbe;
+import com.xpedite.probes.AnchoredProbe;
+import com.xpedite.probes.CallSite;
+import javassist.CannotCompileException;
+import javassist.ClassPool;
+import javassist.CtClass;
+import javassist.CtMethod;
+import javassist.NotFoundException;
+
+import java.io.IOException;
+
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.IllegalClassFormatException;
+import java.security.ProtectionDomain;
+
+public class ClassTransformer implements ClassFileTransformer {
+    AbstractProbe[] probes;
+    public ClassTransformer(AbstractProbe[] probes) {
+        this.probes = probes;
+    }
+
+    @Override
+    public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined,
+                            ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+
+        byte[] bytecode = classfileBuffer;
+        try {
+            ClassPool cPool = ClassPool.getDefault();
+            CtClass ctClass = cPool.get(className);
+            for (AbstractProbe probe: probes) {
+                if (probe.getClassName().equals(className)) {
```

Comment:
This for loop might be a bit hard to read as there're multiple nested if statements. Perhaps it could be refactored to early exit from the loop if the condition doesn't hold? Something like:

```java
for (AbstractProbe probe: probes) {
  if (!probe.getClassName().equals(className))
    continue;
  CtMethod ctClassMethod = ctClass.getDeclaredMethod(probe.getMethodName());
  const bool classNameMatch = ctClassMethod.getDeclaringClass().getName().equals(className);
  const bool methodNameMatch = ctClassMethod.getName().equals(probe.getMethodName());
  const bool foundRightMethod = classNameMatch  && methodNameMatch;
  if (!foundRightMethod)
    continue;
  CallSite[] callSites = probe.getCallSites();
  if (callSites.length > 1) {
    branch1();
  } else {
    branch2();
  }
}
```

Reply:
Done